src/service/position-observer/position-observer-entry.js

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use baseInstance file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  layoutRectEquals,
+  layoutRectsOverlap,
+  layoutRectsRelativePos,
+} from '../../layout-rect';
+import {
+  PositionObserverFidelity,
+  LOW_FIDELITY_FRAME_COUNT,
+} from './position-observer-fidelity';
+import {dev} from '../../log';
+
+/**
+ * TODO (@zhouyx): rename relativePos to relativePositions
+ * The positionObserver returned position value which includes the position rect
+ * relative to viewport. And viewport rect which always has top 0, left 0, and
+ * viewport width and height.
+ * @typedef {{
+ *  positionRect: ?../../layout-rect.LayoutRectDef,
+ *  viewportRect: !../../layout-rect.LayoutRectDef,
+ *  relativePos: string,
+ * }}
+ */
+export let PositionInViewportEntryDef;
+
+export class PositionObserverEntry {
+  /**
+   *
+   * @param {!Element} element
+   * @param {!PositionObserverFidelity} fidelity
+   * @param {!function(?PositionInViewportEntryDef)} handler
+   */
+  constructor(element, fidelity, handler) {
+    /** @const {!Element} */
+    this.element = element;
+
+    /** @const {!function(?PositionInViewportEntryDef)} */
+    this.handler_ = handler;
+
+    /** @type {!PositionObserverFidelity} */
+    this.fidelity = fidelity;
+
+    /** @type {number} */
+    this.turn = (fidelity == PositionObserverFidelity.LOW) ?
+        Math.floor(Math.random() * LOW_FIDELITY_FRAME_COUNT) : 0;
+
+    /** @type {?PositionInViewportEntryDef} */
+    this.prevPosition_ = null;
+  }
+
+  /**
+   * Call to update entry position
+   * @param {!PositionInViewportEntryDef} position
+   */
+  trigger(position) {
+
+    const prevPos = this.prevPosition_ ;
+    if (prevPos
+        && layoutRectEquals(prevPos.positionRect, position.positionRect)
+        && layoutRectEquals(prevPos.viewportRect, position.viewportRect)) {
+      // position didn't change, do nothing.
+      return;
+    }
+
+    dev().assert(position.positionRect);
+    const positionRect =
+        /** @type {!../../layout-rect.LayoutRectDef} */ (position.positionRect);
+    // Add the relative position of the element to its viewport
+    position.relativePos = layoutRectsRelativePos(positionRect,
```

Comment:
Why is this being calculated here?

Reply:
Hmmm that's how we current define the return value of PositionObserverCallback 
```
{
  viewportRect,
  positionRect,
  relativePos,
}
```
Because we will be setting `positionRect` to null when element gets out of viewport, we need to calculate its relativePosition to viewport (above/below) here, handler won't have enough data to calculate it.